### DISCUSSION TITLE
Per-target execution platform constraints for exec groups

### DISCUSSION THREAD

**Discussion by fmeum** [bazelbuild/bazel#23245] [Design proposals]
The official discussion of the ["Per-target execution platform constraints for exec groups"](https://docs.google.com/document/d/1u4zP5MLU3HOt-qlyiNnlxcu4SqOB5tnm32v02vLnd5U/edit) proposal, which aims to introduce new `*_exec_compatible_with` attributes (most importantly, `test_exec_compatible_with`) to all rules that can be used to influence the selection of execution platforms for actions running in a non-default action group (such as tests).

Feedback is very welcome!

**Comment by katre** [comment]
Thanks for sending this.

Overall, I agree that this is useful functionality and makes a lot of sense. I have two areas I want to focus in on, so I'm going to create two sub-threads, one for each.

**Comment by katre** [comment]
# Syntax

I have a slight preference for your alternate syntax, although it makes expressing both the default and non-default a bit confusing:
```
    exec_compatible_with = {
        "": ["//:has_fast_cpu"],
        "test": ["//:has_gpu"],
    },
```

However, this parallels [how `exec_properties` handles exec group names](https://bazel.build/extending/exec-groups#using-exec-groups-for-exec-properties).

I also am not excited about having semi-dynamic attribute names added to rules, although I acknowledge that since execution groups are a static set the new attributes would not actually change dynamically.

I'm going to ask @comius and @brandjon, as the rule API and Starlark experts: which is cleaner, adding new attributes based on the execution groups, or adding the new `dict[str, list[Label]]` attribute type? (Also @lberki, who commented directly on the doc on this same question).

**Reply by lberki** [reply]
*I* would choose adding the new dict type. I am biased by the fact that our dict types are weirdly asymmetric: we have `str -> label`, `str -> str`, `str -> list[str]` and `label -> str` which seems to be a pretty arbitrary choice of key/value pairs. For completeness, we should have every combination of a given set of key types (say, `str` and `label`) and a given set of value types (say, `str`, `label` and lists thereof). I mean, ideally, we'd allow any reasonable type as a key and a value, but what I'm proposing is much better than the currend weird set of exceptions.

And if we agree that it's better to have a "symmetric" set of dict types, this seems like a prime use case for a `str -> list[label]` one.

Technically, there is a complication that "label" is really more than one type because they can either cause a dependency, not cause a dependency or just cause a loading-phase dependency (see `LabelClass` in `Type.java`), but I would be fine with ignoring that and making labels in dicts always cause dependencies.

**Reply by brandjon** [reply]
> Technically, there is a complication that "label" is really more than one type

AFAIK we don't have any of the exotic label types exposed to Starlark.

If we wanted symmetry / closure, and especially if we wanted to do it with exotic label types, I think we'd have to move to a fancier way of describing attribute types than simply enumerating combinations manually.

Adding new types of attributes has downstream effects on tooling that needs to understand how to read and write them in BUILD files.

**Reply by brandjon** [reply]
(For some reason I thought GH discussions would allow arbitrary thread trees. But it looks like my knowledge of how to communicate online is still stuck in the days of usenet. So responding to John's question here.)

IIUC, the proposal that programmatically adds new attribute names is: "Process some execution-group-related parameter of a `rule()` definition, and based on that, automatically add attribute definitions to its schema following a naming pattern. Then expect those attributes to be present at analysis time and automatically consume them with some special logic."

This sounds a bit hideous to me, simply because of the namespace pollution it entails. Imagine dumping a target in `bazel query` and seeing a few real attributes and a dozen synthetic ones (well, beyond whatever synthetic attributes we already produce -- but at least those don't have programmatically generated names).

If we can address this use case by adding a new attribute type -- a type whose existence is weakly implied by the types we already support -- then I think that's my preference. Even though it may mean we need to define `attr.string_keyed_label_list_dict()`.

**Reply by fmeum** [reply]
Since the new attribute would be native, we _could_ add support for it without exposing `attr.string_list_dict` (or whatever we call it) to Starlark. But @lberki's point still stands that the current selection of Starlark attributes is limited in ways that could be extended without severe implications on tooling (an `attr.label_list_dict` argument in a BUILD file would be a valid value for an `attr.string_list_dict`).

**Reply by lberki** [reply]
Actually, I was only thinking about the internals of Blaze: those also only support the odd assortment of dict-type labels that I listed above. Once that is fixed, plumbing that to Starlark is just a tiny amount of work and as I said above, I think it'd be a net improvement. I'd much rather wait until @comius is back since this sounds like something that he might have an opinion about and it doesn't seem to be urgent enough not to be able to wait a week.

**Reply by comius** [reply]
I think adding `exec_group_compatible_with: dict[str, list[Label]]` is one of the nicest way to do it. Overloading `exec_compatible_with` is also an option - it follows the pattern of overloading a lot of other APIs already have, but then it needs a reserved name for "default" exec group, and the name doesn't point users to exec groups. I'm having hard time to decide between the two, so please pick one.

One alternative is to reuse exec_properties with some string parsing to convert a string to a label list. Cons is string parsing, but a plus is, that's probably adding the least new code / i.e. no new types.

**Comment by katre** [comment]
# Automatic Execution Groups

I think that, even with automatic execution groups, users can currently use `exec_compatible_with` to change the execution platform for the default execution group. Therefore, I think that this would still be useful for users using automatic execution groups, but it's more difficult since those don't have names.

@kotlaja, what do you think? Is this something we should allow users to do, or is it against the ideas behind automatic execution groups.?

**Reply by kotlaja** [reply]
Thanks Fabian for contributing! 

John, I reviewed the doc and I'm currently discussing some details with Fabian (I'll write more here after that). 

By now, my concern is that more exec_groups could be added when no need for it (rule authors could perceive this as suggestion to use exec_groups, which is contrary to the idea of AEGs). As mentioned in [AEGs docs - When should I use a custom exec_group?](https://bazel.build/extending/auto-exec-groups#when-should-use-exec-groups), we suggest rules authors not to use exec_groups if they can use AEGs (in other words, to use exec_groups only when multiple toolchains need to be executed on a single execution platform). 

Still, IIUC this is related to the default `test` exec_group, so this should not a big obstacle. And, I also assume new exec_groups would be added in this case only when a special configuration is really needed (from rule authors perspective). Still, can we actually guarantee that users would use this only when needed? Imagine they specifically add an exec_group "foo" and then use "foo_exec_compatible_with" when not needed (they could have used "exec_compatible_with" from the very beginning). This is more of a philosophical question, but it could still happen, and it is contrary to what AEGs are here for.

Maybe we should restrict this feature only for `test` exec_group since all requests by now are only connected with `test` exec_group. We could try to implement it in a way to be flexible for adding more exec_groups in the future, if needed. Adding it for all exec_groups and creating new rule attribute for each of them sounds like an overkill if we only need it for 1 exec_group. Sure, not many exec_groups exist right now and even less will be needed with the migration of AEGs, but I prefer setting more constraints to feature than overdoing it (harder to remove from Bazel if needed, more non-used rule attributes).

**Reply by katre** [reply]
@kotlaja As an example, let's imagine @fmeum's example where `cc_test` is using automatic exec groups: there is a `cc_test` target that needs to add the `//:has_fast_cpu` constraint for the C++ toolchain's exec group, and add the `//:has_cpu` constraint for the test exec group. It would be useful to be able to specify this directly, rather than relying on the default exec group (and the `exec_compatible_with` attribute) copying into the C++ toolchain's exec group.

Generalizing a bit, it seems likely that there will be future scenarios, even with automatic exec groups, where a user wants to customize the exec constraints for a specific toolchain's exec group. Can we figure out a way to specify this? One option (off the top of my head) is that these automatic exec groups should be named after the toolchain type, which would allow something like (using the dict syntax from [the earlier comment](https://github.com/bazelbuild/bazel/discussions/23245#discussioncomment-10322351)):

```
cc_test(
    ...,
    exec_compatible_with = {
        "@rules_cc//:toolchain_type": ["//:has_fast_cpu"],
        "test": ["//:has_gpu"],
    },
)
```

**Reply by kotlaja** [reply]
> Generalizing a bit, it seems likely that there will be future scenarios, even with automatic exec groups, where a user wants to customize the exec constraints for a specific toolchain's exec group.

If you think this feature will be useful in the future even for AEGs, then I fully agree we should move forward with exposing all exec groups + automatic exec groups to exec_compatible_with. The restriction for only `test` exec_group that I mentioned earlier came from me getting a feeling by now in the team that we should try to be restrictive about new features until we are not sure that will be useful or/and people are requesting it. At first I got impression that this is not a blocker for users but a better way how to specify constraints inside exec groups and that not many users are actually asking for this.

> One option (off the top of my head) is that these automatic exec groups should be named after the toolchain type

This is already the case. In the background, AEG is just a classic exec_group with its toolchain type as a name, so we can easily reuse this as for all other exec groups.

One more thing to note - AEGs also get constraints from exec_compatible_with ([see here](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/analysis/ExecGroupCollection.java#L91)).

> Can we figure out a way to specify this?

With your example, how would we specify a default constraint? Maybe we could use `"all": [...]`. In any case, we should be careful here since an exec_group could be called `all` (or whatever else we decide), but this would be quite rare and, if that happens, users can change the name of that exec_group in case they have it (similar like Fabian mentioned for `*_exec_compatible_with`).

**Reply by fmeum** [reply]
> One more thing to note - AEGs also get constraints from exec_compatible_with ([see here](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/analysis/ExecGroupCollection.java#L91)).

Could you share the rationale for this behavior? Do you perhaps have usage examples you could share? I initially expected automatic exec groups to have the same behavior as explicitly defined ones in this way.

> With your example, how would we specify a default constraint?

I wonder whether this is a common ask. If a default constraint would also affect the `test` exec group, then I would consider wanting to add a constraint too all exec groups quite rare.

**Reply by fmeum** [reply]
> (they could have used "exec_compatible_with" from the very beginning)

Are you referring to the `exec_compatible_with` parameter of `exec_group` here? If so, then it's important to keep in mind that it is very common in Bazel to be using rules one hasn't authored. Being able to specify constraints both on the `rule` definition and on individual targets is more important due to this cultural difference.

**Reply by kotlaja** [reply]
> Could you share the rationale for this behavior? Do you perhaps have usage examples you could share? I initially expected automatic exec groups to have the same behavior as explicitly defined ones in this way.

When AEGs are not enabled, `default-exec-group` has constraints from all toolchains plus from `exec_compatible_with`. What AEGs did is just "splitting" the `default-exec-group` with multiple toolchains. For example, think of an action which uses one toolchain, which by default uses `default-exec-group` when AEGs are not enabled (specific exec_groups are not created). This means that this action will have constraints from `exec_compatible_with` plus **all** toolchains, while it needs only one toolchain. So if AEGs are enabled, one AEG with that needed toolchain can be used for this action. Still, to maintain the same behaviour as before, `exec_compatible_with` is also needed (difference is that constraints from not-used toolchains for selecting exec platform for the actions are ignored). 

> Are you referring to the exec_compatible_with parameter of exec_group here? 

I'm referring to the `exec_compatible_with` parameter of a rule. I was just thinking of an example how this could be used in a wrong way by rule authors and rule users, but this is quite unlucky to happen in my opinion. You can ignore this now since I trust John that there could be a use case to use this feature in the future even for AEGs (reward overweight risk).

**Reply by katre** [reply]
@kotlaja There are two different uses of `exec_compatible_with` here.

One is during rule definition:
```
my_rule = rule(
    ...
    exec_compatible_with = [ ... default exec constraints ],
    exec_groups = {
        "link": exec_group(
            exec_compatible_with = [ link-specific exec constraints ],
        ),
    },
)
```

This syntax doesn't need to change, either with or without automatic exec groups.

The other usage is on targets in BUILD files, and lets users add new exec constraints to exec groups:
```
my_rule(
    name = "my_specific_target",
    ... other attributes ...
    exec_compatible_with = [
        extra exec constraints for the default exec group
    ],
)
```

> When AEGs are not enabled, default-exec-group has constraints from all toolchains plus from exec_compatible_with.

I think this is referring to the **rule parameter** `exec_compatible_with`. This proposal is about the **target attribute** `exec_compatible_with`. Does AEGs have any special handling for the target attribute version?

My assertion is that users who write BUILD files may need to add constraints to automatic exec groups, and that they should probably be able to do this based on the toolchain name, making my above example something like:
```
my_rule(
    name = "my_specific_target",
    ... other attributes ...
    exec_compatible_with = {
        "???": [ extra exec constraints for the default exec group ]
        "@rules_cc//:toolchain_type": [ extra exec constraints for the AEG for this toolchain type ],
    },
```

Does that make sense?

@fmeum: Can you update the proposal to add details on:
1. How this interacts with AEG, and
2. How to specify default exec group in the dict-based version?

Thanks

**Reply by fmeum** [reply]
Will do! 

@katre With the dict-based approach, how would we avoid breaking existing usages of `exec_compatible_with`? We could introduce custom logic that converts a list `["@foo"]` to `{"": ["@foo"]}` or whatever we decide the key for the default exec group to be, but that would require changes to the otherwise generic attribute handling. 

Alternatively, we could introduce a new `exec_group_compatible_with` attribute and have that apply to the non-default exec groups only, thus avoiding the bikeshed of how to reference the default exec group in the dict-based attribute.

**Reply by fmeum** [reply]
I updated the proposal, please take a look.

**Comment by katre** [comment]
Looks good to me, waiting for @comius to return (Thursday this week) and give an opinion.

**Reply by alexeagle** [reply]
Hey John, do you know if anyone has reminded Ivo to take a look here?

**Reply by comius** [reply]
Thanks for a reminder. After 3rd read I finally understood the proposal. 

LGTM. I think user interface is nice and the semantics well defined. It is adding complexity, that will need some learning from the users. I'd suggest we don't forget to update Bazel's documentation pages in this area.

**Comment by katre** [comment]
With @comius's review, go ahead and mark this as Approved. I'm happy to handle the PRs.
